chat-plugins/scavengers.js

```diff
@@ -1,303 +1,603 @@
-/**
-* Scavenger hunts plugin. Only works in a room with the id 'scavengers'.
-* This game shows a hint. Once a player submits the correct answer, they move on to the next hint.
-* You finish upon correctly answering the third hint.
-* In an official hunt, the first three to finish within 60 seconds achieve blitz.
-*/
-
 'use strict';
-const fs = require('fs');
-let scavengers = {
-	status: 'off',
-	blitz: null,
-	numblitzes: 0,
-	hints: null,
-	answers: null,
-	participants: {},
-	finished: [],
-	resultHTML: null,
-	updatinglb: false,
-};
 
-let scavengersRoom = Rooms.get('scavengers');
-if (scavengersRoom) {
-	if (scavengersRoom.plugin) {
-		scavengers = scavengersRoom.plugin;
-	} else {
-		scavengersRoom.plugin = scavengers;
-	}
-}
+const fs = require("fs");
+
+const dataFile = "./config/chat-plugins/scavdata.json";
+const points = [20, 15, 10, 5, 1];
+
+let leaderboard = {};
 
-let scavengersLB = {};
-let dataFile = './config/chat-plugins/scavdata.json';
 fs.readFile(dataFile, 'utf8', function (err, content) {
-	if (err) {
-		scavengersLB = {};
-		return;
-	}
+	if (err) return console.log("ERROR: Unable to load scavenger leaderboard: " + err);
+
 	try {
-		scavengersLB = JSON.parse(content);
-	} catch (e) {
-		scavengersLB = {};
-	}
+		leaderboard = JSON.parse(content);
+	} catch (e) {}
 });
-function addPoints(username, points) {
-	let id = toId(username);
-	if (!id) return;
-	if (id in scavengersLB) {
-		scavengersLB[id].points += points;
-	} else {
-		scavengersLB[id] = {
-			points: points,
-			name: Chat.escapeHTML(username),
-		};
-	}
+
+function addPoints(name, points, noUpdate) {
+	let userid = toId(name);
+
+	if (!userid || userid === 'constructor' || !points) return false;
+
+	if (!leaderboard[userid]) leaderboard[userid] = {name: name, points: 0};
+	leaderboard[userid].points += points;
+	if (!noUpdate) leaderboard[userid].name = name; // always keep the last used name
 }
-function scavWriteData() {
-	fs.writeFile(dataFile, JSON.stringify(scavengersLB));
+
+function writePoints() {
+	fs.writeFile(dataFile, JSON.stringify(leaderboard), err => {
+		if (err) console.log(`ERROR: Failed to write scavengers ladder - ${err}`);
+	});
 }
 
-function getLadder() {
-	return Object.keys(scavengersLB).map(id => [scavengersLB[id].points, scavengersLB[id].name]).sort((a, b) => b[0] - a[0]);
+function visualizeLadder(userid) {
+	userid = toId(userid);
```

Comment:
??

don't pass something to a function as ``userid`` if it isn't already a userid